### PR TITLE
Use classes in place of callables for resolvers

### PR DESCRIPTION
**Pull request by alexarice**
This implements something halfway between what I originally had in #3456 and what @superlopuh suggested in https://github.com/xdslproject/xdsl/pull/3456#issuecomment-2479586661 . I think this a strict improvement over the original which uses callables and lambda functions and will merge it into that PR if people agree

### CODE REVIEW

File: xdsl/irdl/constraints.py [hunk 1]
```diff
@@ -55,7 +56,14 @@ def update(self, other: ConstraintContext):
 
 _AttributeCovT = TypeVar("_AttributeCovT", bound=Attribute, covariant=True)
 
-ResolveType: TypeAlias = Attribute | Sequence[Attribute]
+ResolveType = Attribute | Sequence[Attribute]
```
Comment (superlopuh) [inline xdsl/irdl/constraints.py:59]: This is inferred but also doesn't hurt, makes sure that the RHS is purely in type land
```suggestion
ResolveType: TypeAlias = Attribute | Sequence[Attribute]
```

File: xdsl/irdl/constraints.py [hunk 2]
```diff
@@ -55,7 +56,14 @@ def update(self, other: ConstraintContext):
...
+
+T = TypeVar("T")
```
Comment (superlopuh) [inline xdsl/irdl/constraints.py:61]: Any reason to make this public?
```suggestion
_T = TypeVar("_T")
```

**Review by superlopuh — APPROVED** [review]
LGTM, except that I'd probably move the resolver classes into the constraint classes, and call them all `_Resolver`

**Comment by alexarice** [comment]
I originally had this exactly as you said with inner classes, and then got worried that this would redefine the classes each time the outer class was used, and then read some stack exchange threads that said that inner classes were stupid, and so I got cold feet about it. I can change it back easily

**Comment by superlopuh** [comment]
I'd be interested in seeing the arguments against inner classes, I'm quite used to them and like them in general

**Comment by alexarice** [comment]
https://stackoverflow.com/questions/719705/what-is-the-purpose-of-pythons-inner-classes this thread had a few things in the comments though I found the argument questionable

**Pull request merged**